src/java/org/apache/cassandra/index/sai/disk/v1/TermsReader.java

```diff
@@ -0,0 +1,294 @@
+        }
+    }
+
+    // currently only used for testing
+    private class TermsScanner implements TermsIterator
+    {
+        private final long segmentOffset;
+        private final TrieTermsDictionaryReader termsDictionaryReader;
+        private final Iterator<Pair<ByteComparable, Long>> iterator;
+        private final ByteBuffer minTerm, maxTerm;
+        private Pair<ByteComparable, Long> entry;
+
+        private TermsScanner(long segmentOffset)
+        {
+            this.termsDictionaryReader = new TrieTermsDictionaryReader(termDictionaryFile.instantiateRebufferer(), termDictionaryRoot);
+
+            this.minTerm = ByteBuffer.wrap(ByteSourceInverse.readBytes(termsDictionaryReader.getMinTerm().asComparableBytes(ByteComparable.Version.OSS42)));
+            this.maxTerm = ByteBuffer.wrap(ByteSourceInverse.readBytes(termsDictionaryReader.getMaxTerm().asComparableBytes(ByteComparable.Version.OSS42)));
+            this.iterator = termsDictionaryReader.iterator();
+            this.segmentOffset = segmentOffset;
+        }
+
+        @Override
+        @SuppressWarnings({"resource", "RedundantSuppression"})
+        public PostingList postings() throws IOException
+        {
+            assert entry != null;
+            final IndexInput input = IndexFileUtils.instance.openInput(postingsFile);
+            return new OffsetPostingList(segmentOffset, new ScanningPostingsReader(input, new PostingsReader.BlocksSummary(input, entry.right)));
+        }
+
+        @Override
+        public void close()
+        {
+            termsDictionaryReader.close();
+        }
+
+        @Override
+        public ByteBuffer getMinTerm()
+        {
+            return minTerm;
+        }
+
+        @Override
+        public ByteBuffer getMaxTerm()
+        {
+            return maxTerm;
+        }
+
+        @Override
+        public ByteComparable next()
+        {
+            if (iterator.hasNext())
+            {
+                entry = iterator.next();
+                return entry.left;
+            }
+            return null;
+        }
+
+        @Override
+        public boolean hasNext()
+        {
+            return iterator.hasNext();
+        }
+    }
+
+    private static class OffsetPostingList implements PostingList
+    {
+        private final long offset;
+        private final PostingList wrapped;
+
+        OffsetPostingList(long offset, PostingList postingList)
+        {
+            this.offset = offset;
+            this.wrapped = postingList;
+        }
+
+        @Override
+        public long nextPosting() throws IOException
+        {
+            long next = wrapped.nextPosting();
+            if (next == PostingList.END_OF_STREAM)
+                return next;
+            return next + offset;
+        }
+
+        @Override
+        public long size()
+        {
+            return wrapped.size();
+        }
+
+        @Override
+        public long advance(long targetRowID) throws IOException
```

Comment:
Aside: This is basically never used because of `ScanningPostingsReader`?

Reply:
Correct. I've removed the implementation to throw an `UnsupportedOperationException` instead. It is possible that we could make that the default in `PostingList`. I will look at that and see if it makes sense.